structure/heap/binary_heap.go

```diff
@@ -0,0 +1,98 @@
+package heap
+
+/*
+This is an implementation of a Binary heap (or simply heap), is parameterized with:
+T: The type of elements
+I: The type of the cardinality of the binary heap as a vector
+
+P.s. If you want use a primitive type you have to wrap it into a custom type and then implement the Prioritizable interface
+
+Usage example:
+
+	type MyInt32 int32
+
+	func (a MyInt32) PriorTo(b MyInt32) bool {
+		return a < b
+	}
+	func main() {
+		h := heap.NewBinaryHeapFromSlice[MyInt32, int32]([]MyInt32{4, 2})
+		h.Push(3)
+		fmt.Println(h.Pop())
+	}
+*/
```

Comment:
Usage example are presented separately in a Example function inside the test file, please follow the Go standards whenever possible - comments have a tendency to go out of date very quickly without rigorous checks and since we don't have those checks here, an example function is much better than this.